### ISSUE TITLE
gromacs: builds non-deterministic, pick up builder's cpu features

### ISSUE THREAD

**Issue by SomeoneSerge** [NixOS/nixpkgs#270525]
## Issue description

Cf. e.g. https://hydra.nixos.org/build/240516188/nixlog/1 vs https://hydra.nixos.org/build/76614747/log:

```diff
@@ -1,6 +1,6 @@
--- Detected build CPU vendor - AMD
--- Detected build CPU brand - AMD EPYC 7402P 24-Core Processor
--- Detected build CPU family - 23
--- Detected build CPU model - 49
--- Detected build CPU stepping - 0
--- Detected build CPU features - aes amd apic avx avx2 clfsh cmov cx8 cx16 f16c fma htt lahf misalignsse mmx msr nonstop_tsc pclmuldq pdpe1gb popcnt pse rdrnd rdtscp sha sse2 sse3 sse4a sse4.1 sse4.2 ssse3 x2apic
+-- Detected build CPU vendor - Intel
+-- Detected build CPU brand - Intel(R) Xeon(R) CPU E5-2686 v4 @ 2.30GHz
+-- Detected build CPU family - 6
+-- Detected build CPU model - 79
+-- Detected build CPU stepping - 1
+-- Detected build CPU features - aes apic avx avx2 clfsh cmov cx8 cx16 f16c fma htt intel lahf mmx msr pcid pclmuldq pdpe1gb popcnt pse rdrnd rdtscp sse2 sse3 sse4.1 sse4.2 ssse3 tdt x2apic
```

I think we should build and cache gromacs with x86_64-v3&c and with MPI support on, but we should do so in a deterministic manner

CC @sheepforce @markuskowa

**Comment by markuskowa** [comment]
I agree, that it should be built in deterministic manner. Gromacs detects the CPU features here, but does it really build the code with the detected flags? It should accept the SIMD flags (see here https://github.com/NixOS/nixpkgs/blob/627dda7f87a1a6ec2a39be3ac11c009c3f759524/pkgs/applications/science/molecular-dynamics/gromacs/default.nix#L71). Did that break?

However, I think we should not build it with x86_64-v3 support turned on by default. I would be happy if everything would be built with AVX2 per default, but this is not the nixpkgs standards at moment and we should to keep things consistent.

I would rather recommend to use an overlay to build optimized code. E.g. we (@sheepforce ) do that in our overlay: ) do that in ouverlay (https://github.com/Nix-QChem/NixOS-QChem/blob/813b42461681186d483fd74f9973e20c1080017d/overlay.nix#L14).
That way you have much more control the optimizations.


See also CPU flags discussion here: https://discourse.nixos.org/t/pre-rfc-gradual-transition-of-nixos-x86-64-baseline-to-x86-64-v3-with-an-intermediate-step-to-x86-64-v2/35924

**Comment by SomeoneSerge** [comment]
> However, I think we should not build it with x86_64-v3 support turned on by default. I would be happy if everything would be built with AVX2 per default, but this is not the nixpkgs standards at moment and we should to keep things consistent.

I don't think there's anything wrong with enabling extra architectures for individual packages (same as some of the packages built by Hydra include ROCm support), and I suspect that this is one of the very few packages where this might make a difference. We can have a separate attribute for the baseline build too.

> It should accept the SIMD flags

Yes, but the fact that it still tries to inspect the builder is concerning. I'll check diffoscope


> See also CPU flags discussion here: https://discourse.nixos.org/t/pre-rfc-gradual-transition-of-nixos-x86-64-baseline-to-x86-64-v3-with-an-intermediate-step-to-x86-64-v2/35924

Yes, I'm aware of this drama

**Comment by markuskowa** [comment]
> I don't think there's anything wrong with enabling extra architectures for individual packages (same as some of the packages built by Hydra include ROCm support), and I suspect that this is one of the very few packages where this might make a difference. We can have a separate attribute for the baseline build too.

I am sure it would make a significant difference (haven't benchmarked it). The Rocm support should not break the build on a system that does not a have GPU. However, enabling AVX2 compiler flags (i.e. no runtime detection) will break it on a CPU that does not support it. Thus we should keep it line with what is common practice/accepted standard  in nixpkgs. 
There is still an surprising amount of older machines out there that does not support AVX2. Note that VMs also does not always support AVX2 (even though the host CPU does).

**Comment by SomeoneSerge** [comment]
> However, enabling AVX2 compiler flags (i.e. no runtime detection) will break it on a CPU that does not support it

I don't think so, we just add gccarch-whatever to `requiredSystemFeatures` and Nix would choose the appropriate remote builder

**Comment by sheepforce** [comment]
> I don't think so, we just add gccarch-whatever to requiredSystemFeatures and Nix would choose the appropriate remote builder

I mean you can set it via this line https://github.com/NixOS/nixpkgs/blob/nixos-unstable/pkgs/applications/science/molecular-dynamics/gromacs/default.nix#L6C6-L6C6 to whatever level of optimisation suits you.

However, I also think that there is no point in building it by default in Nixpkgs. I agree that it would be nice to have a global flag to optimise packages and ideally this could happen via setting localSystem (https://nixos.wiki/wiki/Build_flags). But I've never got it working as deep in the dependency tree bootstrapping RustC already fails. As long as the RFC for x86-64-v3 isn't accepted yet, we should keep distributing the least common denominator.

**Comment by SomeoneSerge** [comment]
I do see some avx instructions in the diff between the two linked Hydra builds: https://gist.github.com/SomeoneSerge/44c735db023bb22388dced33981c23af

**Comment by sheepforce** [comment]
> I do see some avx instructions in the diff between the two linked Hydra builds: https://gist.github.com/SomeoneSerge/44c735db023bb22388dced33981c23af
> 
> EDIT: sse: https://gist.github.com/SomeoneSerge/f88ffa195b49a3e808bde7feccc45e38

Hm, curious. Any idea why I am not seeing any AVX assembler instructions? :grimacing: Would be very helpful for checking in the future.

But Gromac's Cmake has a lot of hints for autodetection, yeah :/
See here https://github.com/gromacs/gromacs/blob/6021274691c179db96a7afeb5a9ff04d13fbf647/cmake/gmxManageSimd.cmake#L128
and here https://github.com/gromacs/gromacs/blob/6021274691c179db96a7afeb5a9ff04d13fbf647/cmake/gmxDetectSimd.cmake

I've just skimmed over it. I've no idea yet which flag to actually change :smile:

**Comment by markuskowa** [comment]
> I don't think so, we just add gccarch-whatever to `requiredSystemFeatures` and Nix would choose the appropriate remote builder

The builder should not be a big problem. However, it may not run when someone pulls it from the cache and tries to run it. Having a package fail with "illegal instruction" is not very user friendly.

@SomeoneSerge since you are also working a lot on HPC related topics: would you be interested in mini Zoom conference (possibly January or February)? We could discuss what the biggest needs for Nix-HPC are and what possible solutions there are to address these issues. The Nix-HPC community seems to have grown in the last two years.

**Comment by SomeoneSerge** [comment]
> Hm, curious. Any idea why I am not seeing any AVX assembler instructions?

I haven't actually looked into where these instructions were actually found (I compared the entire directory trees). I see that we build GROMACS with something called "HWLOC" enabled. Is this a library for detecting available ISAs at runtime? There's a possibility we're seeing code from some special branches that are never chosen unless these instructions are supported by the host

> The builder should not be a big problem. However, it may not run when someone pulls it from the cache and tries to run it. Having a package fail with "illegal instruction" is not very user friendly.

This is a matter of which users we're trying to optimize the experience for. One of the two groups of users has to use a special flag or an override: either the users with access to older hardware need to know to use a special attribute (say `gromacs-baseline`) or an override, or the users expecting magic performance out of the box (e.g. newcomers trying to evaluate whether Nixpkgs fits their needs) have to.

> would you be interested in mini Zoom conference (possibly January or February)

I'd like that! We could discuss the details over nixos discourse or matrix
